File: Data Analysis 03/java code/ipCaculate.java
package ip;

import java.io.BufferedReader;
import java.io.FileReader;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;

public class ipCaculate {
	public static void main(String[] args) {
		String ipPath = "/Users/fred/Desktop/Data/ip.txt";
		String accessPath = "/Users/fred/Desktop/Data/access.log";
		
		List<IpBean> ipTable = getIp(ipPath);
		List<Long> ipList = getAccessLog(accessPath);
		
		Map<String, Integer> map = new HashMap<>();
		for(Long ip : ipList) {
			String province = getProvince(ip, ipTable);
			int count = map.getOrDefault(province, 0);
			count++;
			map.put(province, count);
		}
		
		//遍历输出
		Set<String> keySet = map.keySet();
		for(String key : keySet) {
			System.out.println(key + ":" + map.get(key));
		}	
	}
	
	//获取ip地址所对应的省份
	public static String getProvince(long ip, List<IpBean> ipTable) {
		int low = 0;
		int high = ipTable.size() -1;
		int mid = 0;
		while(low <= high) {
			mid = (high + low) / 2;
			if(ip >= ipTable.get(mid).getStartDecimalIp() && ip <= ipTable.get(mid).getEndDecimalIp()) {
				return ipTable.get(mid).getProvince();
			}else if(ip < ipTable.get(mid).getStartDecimalIp()){
				high = mid - 1;
			}else {
				low = mid + 1;
			}
		}
		return null;
	}
	
	/**
	 * 获取访问Ip的十进制地址，并放入list
	 * @return
	 */
	public static List<Long> getAccessLog(String accessPath) {
		List<Long> list = new ArrayList<>();
		try (
				BufferedReader br = new BufferedReader(new FileReader(accessPath));
				){
			String line;
			while((line = br.readLine()) != null) {
				String ipOld = line.split("\\|")[1];
				long ip = transformToDecimal(ipOld);
				list.add(ip);
			}
		} catch (Exception e) {
			e.printStackTrace();
		}
		return list;
	}
	
	public static long transformToDecimal(String ip) {
		return (long)(Integer.parseInt(ip.split("\\.")[3]) 
				+ Integer.parseInt(ip.split("\\.")[2]) * 256 
				+ Integer.parseInt(ip.split("\\.")[1]) * Math.pow(256, 2) 
				+ Integer.parseInt(ip.split("\\.")[0]) * Math.pow(256, 3));
	}
	
	/**
	 * 将Ip对应表导入list，以便于之后对照查找
	 * @return
	 */
	public static List<IpBean> getIp(String ipPath) {
		List<IpBean> list = new ArrayList<>();
		try (
				BufferedReader br = new BufferedReader(new FileReader(ipPath));
				){
			String line;
			while((line = br.readLine()) != null) {
				String startIp = line.split("\\|")[0];
				String endIp = line.split("\\|")[1];
				long startDecimalIp = Long.parseLong(line.split("\\|")[2]);
				long endDecimalIp = Long.parseLong(line.split("\\|")[3]);
				String province = line.split("\\|")[6];
				String vendor = line.split("\\|")[9];
				list.add(new IpBean(startIp, endIp, startDecimalIp, endDecimalIp, province, vendor));
			}
		} catch (Exception e) {
			e.printStackTrace();
		}
		return list;
	}
	
}

File: Data Analysis 03/python code/ip_caculate.py
#获取access.log中的ip地址信息
def getIpList():
    ip_list = []
    with open('Data/access.log', 'rb') as f:
        lines = [l.decode('utf8', 'ignore') for l in f.readlines()]
        for line in lines:
            #获取IP
            ip_list.append(line.split('|')[1])
    return ip_list

#将access.log中的ip地址转换成十进制ip
def transformToDecimal(ipList):
    ip_decimal_list = []
    for ip in ipList:
        #print(int(ip.split('.')[0]) * pow(256, 3) + int(ip.split('.')[1]) * pow(256, 2) + int(ip.split('.')[2]) * 256 + int(ip.split('.')[3]))
        ip_decimal_list.append(int(ip.split('.')[0]) * pow(256, 3) + int(ip.split('.')[1]) * pow(256, 2) + int(ip.split('.')[2]) * 256 + int(ip.split('.')[3]))
    return ip_decimal_list

#获取ip.txt中ip段的起始地址及对应的省份
def getIpTable():
    ip_table = []
    with open('Data/ip.txt') as f:
        for line in f.readlines():
            ip_info = []
            ip_info.append(line.split('|')[2])
            ip_info.append(line.split('|')[3])
            ip_info.append(line.split('|')[6])

            ip_table.append(ip_info)
    return ip_table

def getResult(ip_decimal_list, ip_table):
    result = {}
    for ip in ip_decimal_list:
        low = 0
        high = len(ip_table) - 1
        while (low <= high):
            mid = int((low + high) / 2)
            if ((int(ip) >= int(ip_table[mid][0])) & (int(ip) <= int(ip_table[mid][1]))):
                count = result.get(ip_table[mid][2], 0)
                count += 1
                result[ip_table[mid][2]] = count
                break
            elif (int(ip) < int(ip_table[mid][0])):
                high = mid - 1
            else:
                low = mid + 1
    return result

if __name__ == '__main__':

    ipList = getIpList()
    ip_decimal_list = transformToDecimal(ipList)
    ip_table = getIpTable()
    result = getResult(ip_decimal_list, ip_table)

    print(result)